File: PR1/PR1_Ej3/Cocinero.py
from HamburguesaBuilder import HamburguesaBuilder
from Subject import Subject
from Pedido import Pedido
class Cocinero(Subject):
    def __init__(self, builder):
        self._builder = builder

        self.status = "Tomando pedido"
        self.observers = []
        self.pedidoActual = Pedido()

    def build_hamburguesa(self):
        self.status = "Cocinando"

        self._builder.create_new_hamburguesa()
        self._builder.aniadePan()
        self._builder.aniadeLechuga()
        self._builder.aniadeTomate()
        if hasattr(self._builder, 'aniadeQuesoCabra'):
            self._builder.aniadeQuesoCabra()
        self._builder.aniadeCebolla()
        self._builder.aniadePepinillos()
        self._builder.aniadeBacon()
        self._builder.aniadeCarne()

        self.pedidoActual.aniadeHamburguesa(self._builder.hamburguesa)
        self.status = "Tomando pedido"

    def cambiaReceta(self, builder):
        self._builder = builder

    def attach(self, observer):
        self.observers.append(observer)

    def detach(self, observer):
        self.observers.remove(observer)

    def notify(self):
        for notified in self.observers:
            notified.update(self.pedidoActual)
        self.pedidoActual = Pedido()

File: PR2/pantalla_pedidos_hamburgueseria/lib/Cocinero.dart
import 'HamburguesaNormalBuilder.dart';
import 'ObservadorPedido.dart';
import 'Subject.dart';
import 'Pedido.dart';
import 'HamburguesaBuilder.dart';

class Cocinero implements Subject {
  late HamburguesaBuilder _builder;
  String status = "Tomando pedido";
  List<ObservadorPedido> observers = [];
  Pedido pedidoActual = Pedido();

  Cocinero(HamburguesaBuilder builder) {
    this._builder = builder;
  }

  void buildHamburguesa() {
    status = "Cocinando";

    _builder.aniadePan();
    _builder.aniadeLechuga();
    _builder.aniadeTomate();
    if (_builder is HamburguesaNormalBuilder) {
      (_builder as HamburguesaNormalBuilder).aniadeQuesoCabra();
    }
    _builder.aniadeCebolla();
    _builder.aniadePepinillos();
    _builder.aniadeBacon();
    _builder.aniadeCarne();
    _builder.aniadePrecio();

    pedidoActual.aniadeHamburguesa(_builder.hamburguesa);
    status = "Tomando pedido";
  }

  void cambiaReceta(HamburguesaBuilder builder) {
    _builder = builder;
  }

  @override
  void attach(ObservadorPedido observer) {
    observers.add(observer);
  }

  @override
  void detach(ObservadorPedido observer) {
    observers.remove(observer);
  }

  @override
  void notify() {
    for (var notified in observers) {
      notified.update(pedidoActual);
    }
    pedidoActual = Pedido();
  }
}